utility/servicer/module.go

```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
 		return nil, fmt.Errorf("Error admitting relay: %w", err)
 	}
 
-	// TODO: implement Persist Relay
-	// TODO: implement execution
-	// TODO: implement state maintenance
-	// TODO: validate the response from the node?
-	// TODO: (QUESTION) Should we persist SignedRPC?
-	return nil, nil
+	response, err := s.executeRelay(relay)
+	if err != nil {
+		return nil, fmt.Errorf("Error executing relay: %w", err)
+	}
+
+	// TODO(M6): Look into data integrity checks and response validation.
+
+	relayDigest, relayReqResBytes, shouldStore, err := s.isRelayVolumeApplicable(relay, response)
+	if err != nil {
+		return nil, fmt.Errorf("Error calculating relay service digest: %w", err)
+	}
+	if !shouldStore {
+		return response, nil
+	}
+
+	session, err := s.getSession(relay)
+	if err != nil {
+		return nil, err
+	}
+
+	localCtx, err := s.GetBus().GetPersistenceModule().NewLocalContext()
+	if err != nil {
+		return nil, fmt.Errorf("Error getting a local context to update token usage for application %s: %w", relay.Meta.ApplicationAddress, err)
+	}
+
+	if err := localCtx.StoreServiceRelay(session, relay.Meta.ApplicationAddress, relayDigest, relayReqResBytes); err != nil {
+		return nil, fmt.Errorf("Error recording service proof for application %s: %w", relay.Meta.ApplicationAddress, err)
+	}
+
+	if err := localCtx.Release(); err != nil {
+		s.logger.Warn().Err(err).Msg("failed to release local context")
+	}
+
+	return response, nil
+}
+
+// isRelayVolumeApplicable returns:
+//  1. The signed digest of a relay/response pair,
+//  2. Whether there was a collision for the specific chain (i.e. should the service proof be stored for claiming later)
+func (s *servicer) isRelayVolumeApplicable(relay *coreTypes.Relay, response *coreTypes.RelayResponse) (digest, serializedRelayRes []byte, collides bool, err error) {
```

Comment:
Thanks, renamed to `doesCollide` to stay consistent with my comment above. But this may need to be renamed depending on your answer to the replay vs. legit-but-not reward-eligible relays.